### PR TITLE
Slicing 

### PR DESCRIPTION
**Pull request by mrocklin**
This is @nevermindewe s work.  Creating a PR for comments.

### CODE REVIEW

File: dask/array.py [hunk 1]
```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
     if arrays[0].ndim <= axis:
         arrays = [a[None, ...] for a in arrays]
     return np.concatenate(arrays, axis=axis)
+
+
+def dask_1d_slice(out_name, in_name, slice_spec, shape, blockshape):
```
Comment (mrocklin) [inline dask/array.py]: Do we still use this function?  If not should remove it.

File: dask/array.py [hunk 2]
```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
...
+    a dictionary where the keys are the integer indexes of the blocks that should be sliced
+      and the values are the slices
+
+
+    >>> _slice_1d(100, 20, slice(0, 35))
+    {0: slice(0, 20, None), 1: slice(0, 15, None)}
+    >>> _slice_1d(100, 20, slice(10,15))
+    {0: slice(10, 15, None)}
+    >>> _slice_1d(100, 20, slice(40, 100))
+    {2: slice(0, 20, None), 3: slice(0, 20, None), 4: slice(0, 20, None)}
```
Comment (mrocklin) [inline dask/array.py]: Also have an example with a non-slice index?

File: dask/array.py [hunk 3]
```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
...
+        dask[(out_name, block_num)] = (operator.getitem, (in_name, leftmost_block+block_num), slice(0, blockshape[0]))
+
+    return dask
+
+
+
+def _slice_1d(dim_shape, blocksize, index):
+    """
+    Parameters
+    ----
```
Comment (mrocklin) [inline dask/array.py]: Style nit, should extend `----` further to cover all of the text `Parameters`.  Same as below.

File: dask/array.py [hunk 4]
```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
...
+            
+    for block_num in range(1, num_blocks):
+        dask[(out_name, block_num)] = (operator.getitem, (in_name, leftmost_block+block_num), slice(0, blockshape[0]))
+
+    return dask
+
+
+
+def _slice_1d(dim_shape, blocksize, index):
+    """
```
Comment (mrocklin) [inline dask/array.py]: Usually top line has a one a one line description, followed by empty line, followed by some prose (if applicable) followed by empty line, followed by sections as you have here.

File: dask/array.py [hunk 5]
```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
...
+    num_blocks = int(ceil(float(dim_shape)/blocksize))
+    
+    if index == Ellipsis or index == slice(None, None, None):
+        return {i: index for i in range(num_blocks)}
+
+    if isinstance(index, int):
+        #integer math here is ok.
+        if index >= dim_shape:
+            raise IndexError("index %s is out of bounds for shape %s" % (index, dim_shape))
+        return {index/blocksize : index % blocksize}
```
Comment (mrocklin) [inline dask/array.py]: Assuming integer division will probably fail in python 3.  Might have to use `//` or explicit call to `math.floor`

File: dask/array.py [hunk 6]
```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
...
+    elif isinstance(index, slice):
+        #How do we deal with a step being bigger than the block?
+        #How do we deal with a negative slice?
+        start = index.start or 0
+        stop = index.stop or dim_shape
+        step = index.step
+        start_block = start/blocksize
+        stop_block = int(ceil(stop/blocksize))+1
+
+        res = [_block_slice_step_start(blocksize, block, start, stop, step) for block in range(0, stop_block)]
```
Comment (mrocklin) [inline dask/array.py]: 80 character lines if possible.

File: dask/array.py [hunk 7]
```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
...
+    - if step > blocksize, we are doing basic indexing into the blocks and may end up skipping a block
+    - negative stepping is UNHANDLED as of 2015-01-16
+
+    >>> _block_slice_step_start(20, 0, 10, 20, None)
+    slice(10, 20, None)
+    >>> _block_slice_step_start(20, 0, 20, 50, None) #The starting index falls outside the block, so it returns None
+    >>> _block_slice_step_start(20, 1, 10, 50, None) #The starting index falls in block 0, so this block's slice starts at index 0
+    slice(0, 20, None)
+    >>> _block_slice_step_start(20, 1, 10, 20, None) #The stopping index falls before block 1 so it returns None
+    >>> _block_slice_step_start(20, 0, 3, 100, 25)
```
Comment (mrocklin) [inline dask/array.py]: Hrm, it looks like in the case of uneven step sizes some blocks might be indexed with slices while some might be indexed with integers.  These don't return the same kind of thing when passed to numpy arrays.

```Python
In [4]: x[3:20:25]
Out[4]: array([ 1.])

In [5]: x[3]
Out[5]: 1.0
```

So it might be wise to return a slice regardless.

File: dask/array.py [hunk 8]
```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
...
+    #  that the given step occurs at. Otherwise, return -1
+    r = range(start, blocksize*(blocknum+1), step)
+    for idx in r:
+        if _index_in_block(idx, blocknum, blocksize):
+            return idx
+    return -1
+            
+
+def _abs_index_to_rel_index(index, blocksize):
+    return index % blocksize
```
Comment (mrocklin) [inline dask/array.py]: can we just inline the `%`?

File: dask/tests/test_slicing.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+def test_dask_slice_2d():
+    expected = {('y', 0, 0): (operator.getitem,
+                               ('x', 0, 0),
+                               (slice(13, 20, 2), slice(10, 20, None))),
+                 ('y', 0, 1): (operator.getitem,
+                               ('x', 0, 1),
+                               (slice(13, 20, 2), slice(0, 20, None))),
+                 ('y', 0, 2): (operator.getitem,
+                               ('x', 0, 2),
+                               (slice(13, 20, 2), slice(0, 5, None)))}
```
Comment (mrocklin) [inline dask/tests/test_slicing.py]: Nice tests.  Could also use tests with integers, negative numbers, etc... Although maybe these are handled by the unit-tests on the other functions.
